### DISCUSSION TITLE
Whether support FMI

### DISCUSSION THREAD

**Discussion by richman2024** [idaholab/moose#25828] [Q&A General] [answered]
Hi,

Does MOOSE support FMI(Functional Mock-up Interface)? 
If it does, could you please share a guidline or some instances to me? If not, is there any interface or method for co-simulation with other tools like OpenModeilca, FEniCS and etc. 

Thank you!

**Comment by GiudGiud** [comment, marked as answer]
Hello

No we don't. 
But the capabilities offered by FMI should be easy to replicate. 

You can wrap your moose solves in a Python script and read the post processing data output to file. FMI only supports scalar values so you don't have to worry about fields. 

Guillaume

**Reply by richman2024** [reply]
Hi Guillaume,

Thank you for you quick answer!
I don't know how to wrap MOOSE slover in a Python script, and FMU make of '.dll' or '.c' file, is Python script availible? Could you please describe the process more detail (a example should be better)? Thanks a lot!

**Reply by GiudGiud** [reply]
ah so if you are looking for a FMU we don't provide that either and that's harder to do for now. FMI is easier to replicate without following the standard. You ll have to manually wrap around MOOSE using the python `os` or `subprocess` modules. 

An example would be, once you have compiled the heat transfer module for example:
```
import subprocess
import csv

# Run the simulation
subprocess.call("heat_transfer-opt -i diffusion.i", shell=True)

# Parse the output file
import csv
with open('diffusion_out.csv') as csvfile:
    reader = csv.reader(csvfile, delimiter=',')
    for row in reader:
        print(', '.join(row))
```

**Reply by richman2024** [reply]
Thank you! I want to wrap it as FMU so that make MOOSE co-simulation with other electromagnetic simulation tools (seems not found this module in MOOSE) like OpenModelica else. 
I'll try this way, it's like run MOOSE by python script and get the results after the simulation finished. Can I get data of each time-steps and set parameters during the simulation progress dynamicly?

**Reply by GiudGiud** [reply]
We have an EM module, it just only does waves for now.
The csv file will have the data for each time step.
To change parameters dynamically, there's two options:
- run each input with a single time step and use the restart capabilities
- try to figure out the web server. I dont have experience on that

**Reply by richman2024** [reply]
Thank you a lot!
Where can I find the guidline for csv file's settings to get the data what I want? For example, I want to get the temperature value at some particular points by time steps in 3D (mesh type is exodus mesh file).

**Reply by GiudGiud** [reply]
use the postprocessors for that. A PointValue will do what you want
https://mooseframework.inl.gov/syntax/Postprocessors/index.html

**Reply by richman2024** [reply]
Thank you!
Is the value of `point = ` means 'x y z' of coordinate. When I set it as `point = '5 5 5'`, I got below error: 
`The following error occurred in the object "test_point", of type "PointValue".

No element located at (x,y,z)=(       5,        5,        5) in PointValue Postprocessor named: test_point
`
Only when I set it as '0 0 0' then it can runing correct.
I use exodus mesh file instead of generate mesh by MOOSE. How can I get the value of a particular point or a plane? 
Besides, I just get 1 step value by this way. How can I get more value by time steps?

**Reply by GiudGiud** [reply]
if you run your simulation with --mesh-only what does it report for the mesh bounding box?

**Reply by richman2024** [reply]
Part of its report:
`Mesh Information:
elem_dimensions()={3}
spatial_dimension()=3
n_nodes()=52815
n_local_nodes()=13452
n_elem()=34746
n_local_elem()=8688
n_active_elem()=34746
n_subdomains()=77
n_elemsets()=0
n_partitions()=4
n_processors()=4
n_threads()=1
processor_id()=0
is_prepared()=true
is_replicated()=true

Mesh Bounding Box:
Minimum: (x,y,z)=( -0.3, -0.3, -0.01)
Maximum: (x,y,z)=( 0.3, 0.3, 0.004)
Delta: (x,y,z)=( 0.6, 0.6, 0.014)

Mesh Element Type(s):
HEX8

Mesh Nodesets:
Nodeset 1, 32238 nodes
Bounding box minimum: (x,y,z)=( -0.3, -0.3, 0)
Bounding box maximum: (x,y,z)=( 0.3, 0.3, 0.004)
Bounding box delta: (x,y,z)=( 0.6, 0.6, 0.004)
............................
Mesh Sidesets:
Sideset 1, 31470 sides (QUAD4), 30990 elems (HEX8), 32231 nodes
Side volume: 0.687315
Bounding box minimum: (x,y,z)=( -0.3, -0.3, 0)
Bounding box maximum: (x,y,z)=( 0.3, 0.3, 0.004)
Bounding box delta: (x,y,z)=( 0.6, 0.6, 0.004)
Sideset 2 (r1_wall), 132 sides (QUAD4), 80 elems (HEX8), 155 nodes
Side volume: 0.003444
Bounding box minimum: (x,y,z)=( 0.1815, -0.236, -0.01)
Bounding box maximum: (x,y,z)=( 0.2685, -0.214, 0)
Bounding box delta: (x,y,z)=( 0.087, 0.022, 0.01)
...................................
Mesh Edgesets:
None

Mesh Subdomains:
Subdomain 1: 32900 elems (HEX8, 32900 active), 50073 active nodes
Volume: 0.00144
Bounding box minimum: (x,y,z)=( -0.3, -0.3, 0)
Bounding box maximum: (x,y,z)=( 0.3, 0.3, 0.004)
Bounding box delta: (x,y,z)=( 0.6, 0.6, 0.004)
..............................................`

**Reply by GiudGiud** [reply]
your mesh only extends to Bounding box maximum: (x,y,z)=( 0.3, 0.3, 0.004)

so5,5,5 is too far.
You need to check the coordinates

**Reply by richman2024** [reply]
Oh you are right! It worked, The unit is "m".
There are 2 more questions:
1. How to just get data from a plane?
2. How to get data by more steps? It only get the value of one step which is the final result.

**Reply by GiudGiud** [reply]
a single value for an entire plane? or a grid of values in a plane?

the value is computed at every time step (well that s the default, you can change it). If you look at the csv file you can see more than 1 step value

**Reply by richman2024** [reply]
1. A grid of values in a plane.
2. How to change the setting to get each time steps value? 
My setting as
`[test_point]
    type = PointValue
    point = '0.1 0.1 0.002'
    variable = T
    execute_on = 'initial'
  []`
Then the data in csv is 
`time, test_point
0, 293.89
1, 293.89`

**Reply by GiudGiud** [reply]
1) You could specify the entire grid in the parameter
https://mooseframework.inl.gov/source/vectorpostprocessors/PointValueSampler.html

We dont have a PlaneValueSampler. We could use one, or we could work with the Positions system instead. 

2) change that
execute_on = 'initial'
to 
execute_on = 'timestep_end'

**Reply by richman2024** [reply]
Thank you!
I change the setting to `execute_on = 'timestep_end'`, then I get
time, test_point
0, 0
1, 296.68206308935
it's still showing 1 step value, but that's enough. Thank you for take your many time!

**Reply by aaelmeli** [reply]
@GiudGiud does `points=` in the `PointValueSampler` accept function or can pass the grid as csv? I am asking because writing the coordinates of many points will be a difficult task.

**Reply by GiudGiud** [reply]
Hello

Ok let s use the Positions system instead for that.
You will find numerous ways to specify points there:
https://mooseframework.inl.gov/syntax/Positions/index.html
If you have more needs we can look at implementing them, it s a new system

Then to sample you will need to use this PositionsFunctorSampler object, which will be merged soon (like tomorrow maybe)
https://github.com/idaholab/moose/pull/25872

Guillaume

**Reply by GiudGiud** [reply]
We're delayed on merging this due to ongoing deployment work. Still likely for this week, and please feel free to check it out locally

**Comment by chrbertsch** [comment]
> FMI only supports scalar values so you don't have to worry about fields.

This is true for FMI 1.0 and 2.0, not for 3.0, see https://fmi-standard.org/docs/3.0.1/

**Reply by GiudGiud** [reply]
Thanks for sharing the link!
do you mind pasting a section number?

**Comment by chrbertsch** [comment]
https://fmi-standard.org/docs/3.0.1/#fmu-variables

**Reply by GiudGiud** [reply]
Thank you!

So to do a "Transfer" in the MOOSE sense ( https://mooseframework.inl.gov/syntax/Transfers/index.html ) we would send 4 variables (one for each of x, y, z and the variable) and the either the receiving FMU or some wrapping code in between would translate that to the potentially non-matching (because the number of cells are different, coarser mesh for example) receiving code?
Or is the conversion (between different representations of the same field, again due to non-matching geometries/meshes) codified in the standard as well?

from what I see in the standard with the get and set routines, I would expect the code in between to do the conversion. 
Or be limited to fully-matching field representations (a CopyTransfer in MOOSE-speak)

Looking around it seems people are planning to codify the representation of variables (fields?) here
https://github.com/modelica/fmi-ls-struct
and their communication here
https://github.com/modelica/fmi-ls-bus

am I correct?

**Comment by chrbertsch** [comment]
Hm. I do not know anything obout Moose, so the "translation" is difficult for me.
The FMI Interface does not know anything about meshes etc. The variable are typically used as "signals", not as spacially distributed fields in the multy-physics sense. 

But your could give new meaning to certain array variables in a layered standard on top of FMI (see https://fmi-standard.org/docs/3.0.1/#VersioningLayered and https://www.conftool.com/modelica2023/index.php/Bertsch-Beyond_FMI_-_Towards_New_Applications_with_Layered_Standards-132_a.pdf?page=downloadPaper&filename=Bertsch-Beyond_FMI_-_Towards_New_Applications_with_Layered_Standards-132_a.pdf&form_id=132&form_version=final  for the concept of layered standards) for your use cases - this is what is done for in fmi-ls-bus and fmi-ls-struct.

Before you go the a spatially distributed multi-phyiscs-extension of FMI, I think exchanging avarage values e.g.  on surfaces with a system simulation tool via signals is the lower hanging fruit. See e.g.  https://www.conftool.com/modelica2023/index.php/Willeke-A_preCICE-FMI_Runner_to_Couple_FMUs_to_PDE-Based_Simulations-175_a.pdf?page=downloadPaper&filename=Willeke-A_preCICE-FMI_Runner_to_Couple_FMUs_to_PDE-Based_Simulations-175_a.pdf&form_id=175&form_version=final

**Comment by burritotukituki** [comment]
> Hello
> 
> No we don't. But the capabilities offered by FMI should be easy to replicate.
> 
> You can wrap your moose solves in a Python script and read the post processing data output to file. FMI only supports scalar values so you don't have to worry about fields.
> 
> Guillaume

Hi,
I am also interested in wrapping MOOSE into a FMI object. So far we are able to do what @GiudGiud suggested above. In our case MOOSE is the master app and Modelica the slave, however the simulation is quite expensive as the Modelica simulation needs to restart in every simulation step. A FMI would be a nice fix to improve performance.

@chrbertsch and @richman2024 are you aware of any FEM or CFD code that has been wrapped into a FMU?
Thanks!

**Reply by GiudGiud** [reply]
they have the list of codes there
https://fmi-standard.org/tools/
there's quite a few

**Reply by yifanm4** [reply]
> > Hello
> > No we don't. But the capabilities offered by FMI should be easy to replicate.
> > You can wrap your moose solves in a Python script and read the post processing data output to file. FMI only supports scalar values so you don't have to worry about fields.
> > Guillaume
> 
> Hi, I am also interested in wrapping MOOSE into a FMI object. So far we are able to do what @GiudGiud suggested above. In our case MOOSE is the master app and Modelica the slave, however the simulation is quite expensive as the Modelica simulation needs to restart in every simulation step. A FMI would be a nice fix to improve performance.
> 
> @chrbertsch and @richman2024 are you aware of any FEM or CFD code that has been wrapped into a FMU? Thanks!

Hi @burritotukituki ,

I am also thinking about getting Modelica to talk with my MOOSE-based application. I understand how to get and read the output csv file from a python script, but I am not familiar with the restart option in MOOSE. Will you be able to give some detailed tutorial on how you use MOOSE as the master app and Modelica as the slave? I think it will serve as a good starting point for me. 

Besides, do you end up succeeding in wrapping MOOSE into a FMI object? If so, can you please give me some suggestions on how it can be done? 

Thank you so much!

**Comment by chrbertsch** [comment]
> @chrbertsch and @richman2024 are you aware of any FEM or CFD code that has been wrapped into a FMU?

You can filter the FMI tools list with "FMI" or "CFD" https://fmi-standard.org/tools/. 
Mostly the import of Functional Mock-up Units is supported by FMI and CFD tools.  
See also  https://www.3ds.com/support/hardware-and-software/simulia-system-information/compatibility/co-simulation/ "Abaqus 2019 and later use the FMI 2.0 standard for communicate with Dymola."

Regarding the export of FMUs many tools can generate reduced-order or surrogate models as FMUs (e.g. Ansys Twin Builder, Optislang ...), 
But there are als tools such as XFlow that support the export of the unreduced model as FMU
